### PR TITLE
[PW-4742] Differentiate partial refunds when doing partial credit memos

### LINKED ISSUE #987: [PW-4742] Adyen partial refund - pspreference don't match and of by one cent refunds
**Describe the bug**
When we have partial refunds based on multiple credit-memo's there is no clear reference which refund belongs to which creditmemo. The reference is only back to the original order.

There are 2 things off here:
- price in cents of by cents with the refund (checked the refund setting, which is based on `refund from first payment first`
- the reference in `adyen_notification` is the order increment id and not the credit memo id

**To Reproduce**
1. Create a partial refund
2. Refund online
3. A log record `REFUND` is returned in `adyen_notification`

**Expected behavior**
There should be a correct relation back to the actual refund.
This could be

**Magento version**
`2.4.1`

**Plugin version**
`6.6.9`

**Screenshots**
Backend setting for refund
image

Related data in `adyen_notification`
image

**Additional context**
We cannot match the `pspreference` with the corresponding credit memo which makes it impossible to do a reconciliation.
Also a second refund doesn't create a extra transaction in the transactions table (this could be a Magento issue)

**Magento insides**
- Credit memos are linked to Orders
- Invoices are linked to Orders
- Payments are linked to Orders
- Transactions are linked to Payments
- No reference exists between Invoice/Credit memo and Transactions/Payments

### PR DESCRIPTION
**Pull request by tnaber**
**The problem**

It is not possible to differentiate between multiple partial credit memos. When multiple partial refunds are made:
- the 'adyen_notification' table will contain the unique pspreferences and the original_reference (after receiving the notification)
- the 'sales_creditmemo' table will contain a transaction_id (which only links to the original_reference), the invoice_id, order_id, and amount details

&#8594; looking at the 'sales_creditmemo' table it is not possible to tell which partial refund entry belongs to which notification in adyen_notification.
For a more detailed explanation see: #987 or look at the screenshots below! 

**The solution**
Checkout this reply for a clear overview of the implemented solution: https://github.com/Adyen/adyen-magento2/issues/987#issuecomment-1018357700
Note: purposely built in a similar fashion to #1295 (partial captures)
Note 2: this PR does not fix an active problem, it aims to allow us and merchants to expand creditmemo functionalities without encountering this problem

Some screenshots for visual support:

adminpanel
image

sales_creditmemo
image

adyen_notification
image

adyen_creditmemo
image

Steps:
- [x] Add database table to upgradeSchema
- [x] Add entity files
- [x] Persist creditmemo's in adyen_creditmemo table
- [x] Create links to differentiate

**Tested scenarios**
- [x] Partial refund manually
- [x] Partial refund for giftcards
- [x] Normal refund

**Fixed issue**:  
Fixes #987

### CODE REVIEW

**Review by AlexandrosMor — APPROVED** [review]

File: Gateway/Http/Client/TransactionRefund.php [hunk 1]
```diff
@@ -30,6 +30,10 @@
  */
 class TransactionRefund implements ClientInterface
 {
+    const REFUND_AMOUNT = 'capture_amount';
```
Comment (Morerice) [inline Gateway/Http/Client/TransactionRefund.php:33]: Could you change the string to `refund_amount`?

File: Api/Data/CreditmemoInterface.php [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    /**
+     * @return mixed
+     */
+    public function getUpdatedAt();
+
+    /**
+     * @param $updatedAt
+     */
+    public function setUpdatedAt($updatedAt);
+}
```
Comment (Morerice) [inline Api/Data/CreditmemoInterface.php:128]: Could you add a new line here?

File: Gateway/Response/PaymentRefundDetailsHandler.php [hunk 3]
```diff
@@ -23,28 +23,54 @@
...
             }
 
             // set pspReference as lastTransId only!
             $payment->setLastTransId($singleResponse['pspReference']);
+
+            $this->creditmemoHelper->createAdyenCreditmemo(
+                $payment,
+                $singleResponse['pspReference'],
+                $singleResponse[TransactionRefund::ORIGINAL_REFERENCE],
+                (int) $singleResponse[TransactionRefund::REFUND_AMOUNT]
```
Comment (Morerice) [inline Gateway/Response/PaymentRefundDetailsHandler.php:72]: Could you use the `intval` function instead of casting?

File: Helper/Creditmemo.php [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        /** @var \Adyen\Payment\Api\Data\OrderPaymentInterface $adyenOrderPayment */
+        $adyenOrderPayment = $this->orderPaymentResourceModel->getOrderPaymentDetails($originalReference, $payment->getEntityId());
+
+        /** @var AdyenCreditmemoModel $adyenCreditmemo */
+        $adyenCreditmemo = $this->adyenCreditmemoFactory->create();
+        $adyenCreditmemo->setPspreference($pspReference);
+        $adyenCreditmemo->setOriginalReference($originalReference);
+        $adyenCreditmemo->setAdyenPaymentOrderId($adyenOrderPayment[\Adyen\Payment\Api\Data\OrderPaymentInterface::ENTITY_ID]);
+        $adyenCreditmemo->setAmount($this->adyenDataHelper->originalAmount($refundAmountInCents, $order->getBaseCurrencyCode()));
+        // TODO: Add credit memo status here
```
Comment (Morerice) [inline Helper/Creditmemo.php:94]: Has this TODO been covered or is it something intended for a future update?

File: Setup/UpgradeSchema.php [hunk 5]
```diff
@@ -616,6 +620,9 @@ public function updateSchemaVersion730(SchemaSetupInterface $setup)
     public function updateSchemaVersion801(SchemaSetupInterface $setup)
```
Comment (Morerice) [inline Setup/UpgradeSchema.php:620]: Could you update the version since we're currently on 8.1.1?

**Pull request closed without merging**